File: schedule/layerlevel.py
from utils.consts    import  *
import layers.layer
#from layers.layer import Layer

##########################################################
class LayerLevel(object):
    #-----------------------------------------------------------------
    def __init__(self, levelNum, initLayers):
        isinstance(initLayers, list)
        assert(initLayers==[] or isinstance(initLayers[0], layers.layer.Layer))
        self.__LevelNum = levelNum
        self.__Layers = initLayers

    #-----------------------------------------------------------------
    def append(self, layer):
        assert(isinstance(layer, layers.layer.Layer))
        self.__Layers.append(layer)

    #-----------------------------------------------------------------
    def gLevelNum(self):
        return self.__LevelNum

    #-----------------------------------------------------------------
    def gLayers(self):
        return iter(self.__Layers)

    #-----------------------------------------------------------------
    def gNumberLayers(self):
        return len(self.__Layers)

    #-----------------------------------------------------------------
    def qDataLevel(self):
        for layer in self.__Layers:
            if layer.gLayerType() != LAYER_TYPE_DATA:
                return False
        return True

File: compiler/kcc/schedule/src/layerlevel.cpp
#include <algorithm>

#include "consts.hpp"
#include "layer.hpp"
#include "layerlevel.hpp"

namespace kcc {
using layers::Layer;

namespace schedule {

LayerLevel::LayerLevel(int levelNum, const std::vector<Layer*>& initLayers)
    : m_LevelNum(levelNum)
    , m_Layers(initLayers)
{}


//--------------------------------------------------------
void
LayerLevel::remove(Layer* layer)
{
    //assert(qContainsLayer(layer));
    auto it = find(m_Layers.begin(), m_Layers.end(), layer);
    assert(it != m_Layers.end());
    auto lastIt = m_Layers.end() - 1;
    std::swap(*it, *lastIt);
    m_Layers.pop_back();
}

//--------------------------------------------------------
void
LayerLevel::append(Layer* layer)
{
    m_Layers.push_back(layer);
}

//--------------------------------------------------------
bool
LayerLevel::qDataLevel() const
{
    for (auto layer : m_Layers) {
        if (!layer->qInputLayer()) {
            return false;
        }
    }
    return true;
}

//--------------------------------------------------------
bool
LayerLevel::qContainsLayer(Layer* layer) const
{
    return std::find(m_Layers.begin(), m_Layers.end(), layer) != m_Layers.end();
}

}}

